### PR TITLE
vvvvvv: init at 2.3.6

### PR DESCRIPTION
**Pull request by SuperSamus**
###### Description of changes

Adds VVVVVV, closing #77519 and superseding #79068.

###### Things done

Also updated tinyxml-2.
Didn't test Darwin at all. If no one can test it, I'll remove it.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dali99** [comment]
How does this address the problems discussed in #79068 ?

**Comment by tpwrules** [comment]
Result of `nixpkgs-review pr 176140` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>backintime</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>backintime-common</li>
    <li>encfs</li>
    <li>tinyxml-2</li>
    <li>vvvvvv</li>
  </ul>
</details>

Plays great too!

**Comment by SuperSamus** [comment]
> How does this address the problems discussed in #79068 ?

To be fair, I looked at it less than I should.
As not a lawyer, and if I go by semantics, the thing that cannot be done is to distribute data.zip. Technically, what's being done is using a script to download data.zip from the official source, so it should be as ok as downloading it manually. (I may be wrong, though.)

**Comment by SuperSamus** [comment]
Should I also add a `vvvvvv-redist` (or similar name) in `all-packages.nix` that overrides `makeAndPlay` to true or would it have no use?

File: pkgs/games/vvvvvv/default.nix [hunk 1]
```diff
@@ -0,0 +1,86 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, fetchurl
+, cmake
+, makeWrapper
+, copyDesktopItems
+, makeDesktopItem
+, pkgconfig
```
Comment (OPNA2608) [inline pkgs/games/vvvvvv/default.nix:9]:
```suggestion
, pkg-config
```

File: pkgs/games/vvvvvv/default.nix [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  sourceRoot = "source/desktop_version";
+  dataZip = fetchurl {
+    url = https://thelettervsixtim.es/makeandplay/data.zip;
+    name = "data.zip";
+    sha256 = "sha256-b6483sBgYtBYJ9QYHEOBU/PqOQBDekTbc7zSl5n+V+A=";
+  };
+
+  cmakeFlags = [ "-DBUNDLE_DEPENDENCIES=OFF" ] ++ lib.optional makeAndPlay "-DMAKEANDPLAY=ON";
+
+  nativeBuildInputs = [ cmake makeWrapper copyDesktopItems pkgconfig ];
```
Comment (OPNA2608) [inline pkgs/games/vvvvvv/default.nix:39]:
```suggestion
  nativeBuildInputs = [ cmake makeWrapper copyDesktopItems pkg-config ];
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
```console
error: Function called without required argument "pkgconfig" at /Users/puna/.cache/nixpkgs-review/pr-176140/nixpkgs/pkgs/games/vvvvvv/default.nix:9, did you mean "pkg-config" or "pkgconfig"?
```

File: pkgs/games/vvvvvv/default.nix (hunk 1 above)
Comment (SuperSamus) [reply pkgs/games/vvvvvv/default.nix]: `pkg-config` isn't actually needed at all, but I messed up with Git and accidentally put it back without noticing (sorry for the mistake).

**Review by OPNA2608 — APPROVED** [review]
Result of `nixpkgs-review pr 176140` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>backintime</li>
    <li>backintime-common</li>
    <li>encfs</li>
    <li>tinyxml-2</li>
    <li>vvvvvv</li>
  </ul>
</details>

Both the full game & the Make and Play edition play fine. Would it make sense to add a top-level entry for the latter & explain its differences with an `optionalString`'d addition to the descriptions?

**Comment by SuperSamus** [comment]
How do I name it? `vvvvvv-makeandplay`? `vvvvvv-redist`?

**Comment by SuperSamus** [comment]
Sorry for bumping this, but:
> How do I name [the Make And Play version in top-level]? `vvvvvv-makeandplay`? `vvvvvvv-make-and-play`? `vvvvvv-redist`?

**Comment by OPNA2608** [comment]
Sorry! I think `vvvvvv-makeandplay` would be fine.

**Comment by SuperSamus** [comment]
Sorry for the bump, is there anything left to do for this to be merged?

**Comment by dali99** [comment]
Personally I'm ~fine with this, I think I preferred the requireFile solution from the earlier solution though. It's a little difficult since I believe since hydra is going to cache the data.zip it will "redistribute" it even when building the unfree version. Its just being distributed _without_ any of the other binaries at all. Even though I think the allowed distribution is only when distributing it with make and play

 but lets give the people involved in the last PR a chance to have some input

@anna328p @leo60228

**Comment by SuperSamus** [comment]
I gave a quick look.
cache.nixos.org doesn't cache unfree-redistributable packages... yet. #83884
You are right, even with the Make And Play version, data.zip is still there (even in [the official download page](https://thelettervsixtim.es/makeandplay/)), so... yeah, it should absolutely not be cached.
It would be possible to prevent it by either making data.zip itself a derivation (so that it could have `meta.license = unfree`), or by simply adding `meta.hydraPlatforms = []` so that both versions will never ever be cached no matter what happens (I prefer the second option).
At this point, though, I think it doesn't make sense to have the Make And Play version top-level...

TL;DR, if no one says anything, I'm gonna add `meta.hydraPlatforms = []` and remove `vvvvvv-makeandplay` from top-level (I'll still leave it as an option).

File: pkgs/games/vvvvvv/default.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    install -m755 VVVVVV $out/bin/${pname}
+
+    mkdir -p $out/share/pixmaps
+    install -D -m644 "$src/desktop_version/icon.ico" "$out/share/pixmaps/VVVVVV.png"
+
+    wrapProgram $out/bin/${pname} --add-flags "-assets $dataZip"
```
Comment (freezeboy) [inline pkgs/games/vvvvvv/default.nix:69]: Shouldn't it be ${dataZip} instead ?

Comment (SuperSamus) [reply pkgs/games/vvvvvv/default.nix:69]: I guess it doesn't have to, because the game opens and works.

**Comment by anna328p** [comment]
> Personally I'm ~fine with this, I think I preferred the requireFile solution from the earlier solution though. It's a little difficult since I believe since hydra is going to cache the data.zip it will "redistribute" it even when building the unfree version. Its just being distributed _without_ any of the other binaries at all. Even though I think the allowed distribution is only when distributing it with make and play.
> 
> If both are marked as not being redistributable I think the problem goes away
> 
> but lets give the people involved in the last PR a chance to have some input
> 
> @anna328p @leo60228

I think a good way to do this is to make data.zip a derivation with two versions, one unfree-redistributable and one unfree (with requireFile maybe?). pass the unfree derivation to the full version of the game, and the unfree-redistributable version to the make and play edition. this means they would have different hashes? i am not sure

**Comment by leo60228** [comment]
While I think using a `licenses.unfree` `fetchurl` for both M&P and the full version is technically fine, I do think it violates the spirit of the license, and it'd probably be preferable to use `fetchurl` for M&P and `requireFile` for the full version. However, without any additional steps, they would have the same hash. Something like what I proposed in https://github.com/NixOS/nixpkgs/pull/79068#issuecomment-581399903 would solve this.

File: pkgs/games/vvvvvv/default.nix [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    install -m755 VVVVVV $out/bin/${pname}
+
+    mkdir -p $out/share/pixmaps
+    install -D -m644 "$src/desktop_version/icon.ico" "$out/share/pixmaps/VVVVVV.png"
```
Comment (SuperSandro2000) [inline pkgs/games/vvvvvv/default.nix:68]:
```suggestion
    install -dD -m644 "$src/desktop_version/icon.ico" "$out/share/pixmaps/VVVVVV.png"
```

File: pkgs/games/vvvvvv/default.nix [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      terminal = false;
+      categories = [ "Game" ];
+    })
+  ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    install -m755 VVVVVV $out/bin/${pname}
```
Comment (SuperSandro2000) [inline pkgs/games/vvvvvv/default.nix:65]:
```suggestion
    install -dm755 VVVVVV $out/bin/${pname}
```

File: pkgs/games/vvvvvv/default.nix (hunk 4 above)
Comment (OPNA2608) [reply pkgs/games/vvvvvv/default.nix:68]:
```
-d, --directory
       treat all arguments as directory names; create all components of the spec‐
       ified directories
```
```
bt1cn@NixTop:/dev/shm
↪ touch test

bt1cn@NixTop:/dev/shm
↪ install -dD -m644 ./test ./examples/test2
install: cannot create directory './test': File exists
```
```suggestion
    install -Dm644 "$src/desktop_version/icon.ico" "$out/share/pixmaps/VVVVVV.png"
```

File: pkgs/games/vvvvvv/default.nix (hunk 5 above)
Comment (OPNA2608) [reply pkgs/games/vvvvvv/default.nix:65]: Same as above.
```suggestion
    install -Dm755 VVVVVV $out/bin/${pname}
```

**Comment by SuperSamus** [comment]
For some reason it didn't build anymore, and I had to add some CMake stuff.
```
cmakeFlags = [
  # the cmake package does not handle absolute CMAKE_INSTALL_INCLUDEDIR correctly
  # (setting it to an absolute path causes include files to go to $out/$out/include,
  #  because the absolute path is interpreted with root at $out).
  "-DCMAKE_INSTALL_INCLUDEDIR=include"
  "-DCMAKE_INSTALL_LIBDIR=lib"
];
```

And:
```
NIX_CFLAGS_COMPILE = "-I${lib.getDev SDL2_mixer}/include/SDL2";
```

File: pkgs/development/libraries/tinyxml-2/default.nix [hunk 6]
```diff
@@ -2,17 +2,25 @@
...
 
   nativeBuildInputs = [ cmake ];
 
+  cmakeFlags = [
+    # the cmake package does not handle absolute CMAKE_INSTALL_INCLUDEDIR correctly
+    # (setting it to an absolute path causes include files to go to $out/$out/include,
+    #  because the absolute path is interpreted with root at $out).
+    "-DCMAKE_INSTALL_INCLUDEDIR=include"
+    "-DCMAKE_INSTALL_LIBDIR=lib"
+  ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/tinyxml-2/default.nix:22]: Ideally this should be fixed upstream but thats also alright.

**Review by SuperSandro2000 — APPROVED** [review]
